requirements-ci.txt

```diff
@@ -1,14 +1,30 @@
+contourpy==1.2.0 ; python_version >= "3.9" and python_full_version < "4.0.0"
+cvxopt @ https://files.pythonhosted.org/packages/10/dc/1c21715e1267ca29f562e4450426d1ff8a7ffcc3e670100cec332a105b95/cvxopt-1.3.2-cp312-cp312-macosx_10_9_x86_64.whl ; sys_platform == "darwin" and platform_machine == "x86_64" and python_version == "3.12"
+cvxopt @ https://files.pythonhosted.org/packages/9f/ad/edce467c24529c536fc9de787546a1c8eca293009383a872b6f638d22eae/cvxopt-1.3.2-cp312-cp312-win_amd64.whl ; python_version == "3.12" and sys_platform == "windows"
+cvxopt @ https://files.pythonhosted.org/packages/c7/17/ee82c745c5bda340a4dd812652c42fb71efd45f663554a10c3ec45f230df/cvxopt-1.3.2-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl ; python_version == "3.12" and sys_platform == "linux"
+cvxopt @ https://files.pythonhosted.org/packages/cd/c8/a04048143d0329ccd36403951746c1a6b5f1fc56c479e5a0a77efb2064b2/cvxopt-1.3.2-cp312-cp312-macosx_11_0_arm64.whl ; sys_platform == "darwin" and platform_machine == "arm64" and python_version == "3.12"
+cvxopt==1.3.2 ; python_full_version >= "3.9.0" and python_version < "3.12"
+cvxpy==1.4.1 ; python_full_version >= "3.9.0" and python_full_version < "4.0.0"
+cycler==0.12.1 ; python_version >= "3.9" and python_full_version < "4.0.0"
+ecos==2.0.12 ; python_full_version >= "3.9.0" and python_full_version < "4.0.0"
+fonttools==4.46.0 ; python_version >= "3.9" and python_full_version < "4.0.0"
+importlib-resources==6.1.1 ; python_version >= "3.9" and python_version < "3.10"
+kiwisolver==1.4.5 ; python_version >= "3.9" and python_full_version < "4.0.0"
+matplotlib==3.8.0 ; python_version >= "3.9" and python_full_version < "4.0.0"
+numpy==1.26.2 ; python_version >= "3.9" and python_full_version < "4.0.0"
+osqp==0.6.3 ; python_full_version >= "3.9.0" and python_full_version < "4.0.0"
+packaging==23.2 ; python_version >= "3.9" and python_full_version < "4.0.0"
+pillow==10.1.0 ; python_version >= "3.9" and python_full_version < "4.0.0"
+pybind11==2.11.1 ; python_full_version >= "3.9.0" and python_full_version < "4.0.0"
+pyparsing==3.1.1 ; python_version >= "3.9" and python_full_version < "4.0.0"
+python-dateutil==2.8.2 ; python_version >= "3.9" and python_full_version < "4.0.0"
+qdldl==0.1.7.post0 ; python_full_version >= "3.9.0" and python_full_version < "4.0.0"
+scipy==1.11.4 ; python_version >= "3.9" and python_full_version < "4.0.0"
+scs==3.2.4.post1 ; python_full_version >= "3.9.0" and python_full_version < "4.0.0"
+setuptools-scm==8.0.4 ; python_version >= "3.9" and python_full_version < "4.0.0"
+setuptools==69.0.3 ; python_version >= "3.9" and python_full_version < "4.0.0"
+six==1.16.0 ; python_version >= "3.9" and python_full_version < "4.0.0"
+tomli==2.0.1 ; python_version >= "3.9" and python_version < "3.11"
+typing-extensions==4.8.0 ; python_version >= "3.9" and python_full_version < "4.0.0"
+zipp==3.17.0 ; python_version >= "3.9" and python_version < "3.10"
```

Comment:
This is pinning packages where we didn't before. Is that necessary?

Reply:
depends what you mean by necessary. Some notes on that:
- as discussed in another comment, this file can be deleted altogether if we don't think people depend on it. CI doesn't use it anymore (in this PR), I just re-generated it to not delete a file that users might have counted on
- the pinning is just how `poetry export` does it, because that's how Poetry itself behaves. You give it acceptable version ranges, it gives you _one_ exact set of compatible versions of each dependency

I don't want to maintain _both_ poetry and pip-style files manually - if we decide that we still have to maintain these manually for some reason, then I have some hesitations moving forward